Sources/Extensions/AppIntents/Widget/Details/WidgetDetailsAppIntentTimelineProvider.swift

```diff
@@ -0,0 +1,136 @@
+    func placeholder(in context: Context) -> WidgetDetailsEntry {
+        .init(
+            upperText: nil, lowerText: nil, detailsText: nil,
+            runAction: false, action: nil
+        )
+    }
+
+    private func entry(for configuration: WidgetDetailsAppIntent, in context: Context) async throws -> Entry {
+        guard Current.servers.all.count > 0 else {
+            Current.Log.error("Failed to fetch data for details widget: No servers exist")
+            throw WidgetDetailsDataError.noServers
+        }
+        
+        let server = configuration.server.getServer() ?? Current.servers.all.first!
+        let api = Current.api(for: server)
+        
+        let upperTemplate = !configuration.upperTemplate.isEmpty ? configuration.upperTemplate : "?"
+        let lowerTemplate = !configuration.lowerTemplate.isEmpty ? configuration.lowerTemplate : "?"
+        let detailsTemplate = !configuration.detailsTemplate.isEmpty ? configuration.detailsTemplate : "?"
+        let template = "\(upperTemplate)|\(lowerTemplate)|\(detailsTemplate)"
+        
+        let result = await withCheckedContinuation { continuation in
+            api.connection.send(.init(type: .rest(.post, "template"), data: ["template": template], shouldRetry: true)) { result in
+                continuation.resume(returning: result)
+            }
+        }
+        
+        var data: HAData?
+        switch (result) {
+        case let .success(resultData):
+            data = resultData
+        case let .failure(error):
+            Current.Log.error("Failed to render template for details widget: \(error)")
+            throw WidgetDetailsDataError.apiError
+        }
+        
+        var renderedTemplate: String?
+        switch (data!) {
+        case let .primitive(response):
+            renderedTemplate = response as? String
+        default:
+            Current.Log.error("Failed to render template for details widget: Bad response data")
+            throw WidgetDetailsDataError.badResponse
+        }
+        
+        let params = renderedTemplate!.split(separator: "|")
+        guard params.count == 3 else {
+            Current.Log.error("Failed to render template for details widget: Wrong length response")
+            throw WidgetDetailsDataError.badResponse
+        }
+        
+        let upperText = String(params[0])
+        let lowerText = String(params[1])
+        let detailsText = String(params[2])
+        return .init(
+            upperText: upperText != "?" ? upperText : nil,
+            lowerText: lowerText != "?" ? lowerText : nil,
+            detailsText: detailsText != "?" ? detailsText : nil,
+            
+            runAction: configuration.runAction,
+            action: configuration.action?.asAction()
+        )
+    }
+}
+
+enum WidgetDetailsDataSource {
+    static var expiration: Measurement<UnitDuration> {
+        .init(value: 2, unit: .hours)
+    }
+    
+    static var fastExpiration: Measurement<UnitDuration> {
```

Comment:
perhaps for this kind of widget that presents info templates we could use the minimum expiration allowed by Apple, I think it is 15 minutes right? (even though the OS may decide to wait for more time of course)

Reply:
Yeah, I wasn't sure, because of the daily limit

Reply:
Each widget has a budget system between 40-70 credits, let's try using 15 minutes (the lowerst possible) and then we increase it later if needed